draft-ietf-masque-connect-udp-listen.md

```diff
@@ -153,11 +163,114 @@ UDP Payload:
 : The unmodified UDP Payload of this proxied UDP packet (referred to as "data
 octets" in {{UDP}}).
 
+# Context ID {#contextid}
+
+A Context ID of value 0 represents no compression. When using context ID 0, the
+client and server must specify the IP version, IP and port of the target per
+datagram, see {{format}}. The client can compress IP and port information using
+non zero Context IDs as defined under {{compression}}.
+Receiving and forwarding from Context ID 0 is enabled by default, but may be
+disabled using {{restrictingips}}.
+
+## Header Compression {#compression}
+
+The client MAY choose to map IP and port information per datagram against the
+Context ID defined in {{contextid}}. In such a case, the client MUST send a
+COMPRESSION_REQUEST capsule (see {{capsulerequestformat}}) with the target
+information (see {{targetmappingformat}}) it wishes to compress and the proxy
+MUST respond with either a COMPRESSION_ASSIGN (see {{capsuleresponseformat}})
+capsule if it accepts the compression request, or a COMPRESSION_REJECT
+(see {{capsulerejectformat}}) if it doesn't wish to support compression
+(For example, due to considerable memory requirements of establishing a list
+of mappings per target per client). If the compression was rejected,
+the client and server MUST continue to use uncompressed, Context ID 0 Datagrams
+for the requested target.
```

Comment:
Can we add text saying that endpoints MUST NOT use a compressed ID until the peer has accepted it?

Reply:
We decided that the client could preemptively send compressed packets, accepting the fact that they may fall through